linera-views/src/value_splitting.rs

```diff
@@ -0,0 +1,378 @@
+        let mut keys_add = Vec::new();
+        let mut n_blocks = Vec::new();
+        for (key, value) in keys.iter().zip(values) {
+            match value {
+                None => {
+                    n_blocks.push(0);
+                    big_values.push(None);
+                }
+                Some(value) => {
+                    let count = Self::read_count_from_value(&value)?;
+                    let big_value = value[4..].to_vec();
+                    for i in 1..count {
+                        let big_key_segment = Self::get_segment_key(key, i)?;
+                        keys_add.push(big_key_segment);
+                    }
+                    n_blocks.push(count);
+                    big_values.push(Some(big_value));
+                }
+            }
+        }
+        if !keys_add.is_empty() {
+            let segments: Vec<Option<Vec<u8>>> = self.client.read_multi_key_bytes(keys_add).await?;
+            let mut pos = 0;
+            for (idx, count) in n_blocks.iter().enumerate() {
+                if count > &1 {
+                    let value: &mut Option<Vec<u8>> = big_values.get_mut(idx).unwrap();
+                    if let Some(ref mut value) = value {
+                        for _ in 1..*count {
+                            let segment: Vec<u8> = segments.get(pos).unwrap().clone().unwrap();
+                            value.extend(segment);
+                            pos += 1;
+                        }
+                    }
+                }
+            }
+        }
+        Ok(big_values)
+    }
+
+    async fn find_keys_by_prefix(&self, key_prefix: &[u8]) -> Result<Self::Keys, Self::Error> {
+        let mut keys = Vec::new();
+        for big_key in self
+            .client
+            .find_keys_by_prefix(key_prefix)
+            .await?
+            .iterator()
+        {
+            let big_key = big_key?;
+            let len = big_key.len();
+            if Self::read_index_from_key(big_key)? == 0 {
+                let key = big_key[0..len - 4].to_vec();
+                keys.push(key);
+            }
+        }
+        Ok(keys)
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let mut key_values = Vec::new();
+        let mut key = Vec::new();
+        let mut big_value = Vec::new();
+        let mut count = None;
+        let mut pos = 0;
+        for result in self
+            .client
+            .find_key_values_by_prefix(key_prefix)
+            .await?
+            .iterator()
+        {
+            let (big_key, value) = result?;
+            let len = big_key.len();
+            let idx = Self::read_index_from_key(big_key)?;
+            if idx == 0 {
+                count = Some(Self::read_count_from_value(value)?);
+                key = big_key[..len - 4].to_vec();
+                big_value.extend(value[4..].to_vec());
+                pos = 1;
```

Comment:
Should we return an error here if `key` and `big_value` aren't empty? Because that means we found an entry in the database that is missing segments?

Reply:
I don't see where those would raise an error.

I mean if `key` and `big_value` aren't empty but we end up in this branch (`idx == 0`), doesn't that mean that the current `key` and `big_value` weren't completed, and are missing a segment?